### PR TITLE
Remove race conditions involving finding available ports

### PR DESCRIPTION
**Pull request by Hixie**
This is an attempt to reland #18488 with less breakage on macOS.

### CODE REVIEW

**Comment by Hixie** [comment]
cc @goderbauer

**Comment by goderbauer** [comment]
Looks like Travis, Cirrus, and Appveyor are unhappy with this.

**Comment by Hixie** [comment]
Sigh. I was expecting the devicelab to go red, not EVERYTHING. :-)

**Comment by amirh** [comment]
Haven't started reviewing yet, but some details about the root cause and the fix in the PR description will help me to review this.

**Comment by Hixie** [comment]
The code used to have places where it looked for an available port, then passed that port to some other code and had that code open the port. That's an inherently racy operation and should never be done. This PR removes all code that did that.

File: dev/devicelab/lib/framework/runner.dart [hunk 1]
```diff
@@ -28,9 +28,8 @@ Future<Map<String, dynamic>> runTask(String taskName, { bool silent = false }) a
   if (!file(taskExecutable).existsSync())
     throw 'Executable Dart file not found: $taskExecutable';
 
-  final int vmServicePort = await findAvailablePort();
   final Process runner = await startProcess(dartBin, <String>[
-    '--enable-vm-service=$vmServicePort',
+    '--enable-vm-service=0',
```
Comment (amirh) [inline dev/devicelab/lib/framework/runner.dart]: What does it do when you give it 0?
Can't find the answer in dart --hepl

File: dev/devicelab/lib/framework/utils.dart [hunk 2]
```diff
@@ -517,13 +502,20 @@ String extractCloudAuthTokenArg(List<String> rawArgs) {
...
+///
+/// The `prefix`, if specified, is a regular expression pattern and must not contain groups.
+///
+/// The `multiLine` flag should be set to true if `line` is actually a buffer of many lines.
+int parseServicePort(String line, {
+  String prefix = 'An Observatory debugger .* is available at: ',
+  bool multiLine = false,
+}) {
+  // e.g. "An Observatory debugger and profiler on ... is available at: http://127.0.0.1:8100/"
+  final RegExp pattern = new RegExp('$prefix(\\S+:(\\d+)/\\S*)\$', multiLine: multiLine);
```
Comment (amirh) [inline dev/devicelab/lib/framework/utils.dart]: This feels pretty fragile, does the Dart team know we depend on this specific line format?
Maybe worth adding a comment where they print this information to emphasize that Flutter depends on it?

File: dev/devicelab/lib/framework/runner.dart (hunk 1 above)
Comment (Hixie) [reply dev/devicelab/lib/framework/runner.dart]: if you follow the dart code, it takes this path:

https://github.com/dart-lang/sdk/blob/607f4f57697a2e27183057eff01ba4400650fdf7/runtime/bin/main_options.cc#L285

https://github.com/dart-lang/sdk/blob/607f4f57697a2e27183057eff01ba4400650fdf7/runtime/bin/main_options.h#L113

https://github.com/dart-lang/sdk/blob/607f4f57697a2e27183057eff01ba4400650fdf7/runtime/bin/main.cc#L476

https://github.com/dart-lang/sdk/blob/607f4f57697a2e27183057eff01ba4400650fdf7/runtime/bin/vmservice_impl.cc#L226

https://github.com/dart-lang/sdk/blob/607f4f57697a2e27183057eff01ba4400650fdf7/runtime/bin/vmservice/vmservice_io.dart#L41

https://github.com/dart-lang/sdk/blob/607f4f57697a2e27183057eff01ba4400650fdf7/runtime/bin/vmservice/server.dart#L352

...which is documented:

https://master-docs-flutter-io.firebaseapp.com/flutter/dart-io/HttpServer/bind.html

...as "If port has the value 0 an ephemeral port will be chosen by the system." which is what we want here.

Comment (amirh) [reply dev/devicelab/lib/framework/runner.dart]: Maybe add a comment saying the 0 will get the system to choose a free port?

**Review by amirh — APPROVED** [review]
LGTM

File: dev/devicelab/lib/framework/utils.dart (hunk 2 above)
Comment (Hixie) [reply dev/devicelab/lib/framework/utils.dart]: We don't really depend on it so much as we'll just need to update the pattern when they change.

In reality, it's quite possible that there's people out there who do depend on it. Any time you change anything, you could break people.

File: dev/devicelab/lib/framework/runner.dart (hunk 1 above)
Comment (Hixie) [reply dev/devicelab/lib/framework/runner.dart]: done

**Comment by Hixie** [comment]
Thanks for the review. Will land on green.

**Pull request merged**